# Guided Programming 4 - Matrix Inversion Using Row Operations

**Mathematical Methods for Chemical Engineers (MTHS1008)**

**Dr Matthew Scase**

In this notebook we will be using what we have learnt about row operations to invert a matrix using Python.

The stages we need to go through are

- Enter the matrix we want to invert
- Create an augmented matrix
- Carry out row operations until we have our inverted matrix.  This will need breaking down into sub-tasks
- Output our inverted matrix
- Check that our output matrix really is the required inverse, i.e., execute some form of error checking

We will being by implementing our steps explicitly and then we can consider whether any of these steps can be refined once we know our algorithm is performing correctly.

So that we can work on a specific example we will invert

$$ \textbf{A} = \left(\begin{array}{rrr} \phantom{-}4 & \phantom{-}2 & -1 \\ 0 & 1 & 3 \\ 3 & 2 & 1 \end{array}\right)$$

There is nothing particularly special about this matrix, I've tried to choose one where the numbers work out reasonably nicely, but other than that it's just a generic matrix.

## Entering the matrix we want to invert

What we know as a 'matrix' in mathematics is known as an 'array' computer programming.

We wish to enter our matrix $\textbf{A}$ into Python.  There are many different ways of doing this.  The method I've chosen here uses the numerical Python module `numpy`.  

We will begin by importing the numerical Python module `numpy`.  To save ourselves writing out "numpy" every time we wish to use one of the functions in the module we can import numpy `as np`.

In [1]:
import numpy as np

Now we can enter our array using the `np.array` function.  The function `np.array` allows us to enter an argument in round brackets, the argument we enter will be our array `A`.  Our array is contained within square brackets.  Each **row** within that array is contained within its own square brackets.

Observe that immediately after writing `np.array` I therefore have to open a set of round brackets to enter my array, then I open a set of square brackets to begin specifying my array, and then I have to open *a further set of square brackets* to begin entering the first row of my array.

`np.array([[`

Each element of the array is separated by a comma.  When I have finished entering my first row I close the square brackets associated with that first row, and then use a comma to indicate I have not finished, then I open another set of square brackets to begin the second row

`np.array([[4, 2, -1], [`
    
When I have finished the final row I close it with a square bracket, then I close my array with a second square bracket, and finally I use a closing round bracket to let `np.array` know I have finished entering the argument.

In [2]:
A = np.array([[4, 2, -1], [0, 1, 3], [3, 2, 1]])
print(A)

[[ 4  2 -1]
 [ 0  1  3]
 [ 3  2  1]]


By using the command `print(A)` I can output the array to the screen to check that it has been entered correctly.  In a final version of the code we could omit this line to improve speed of execution.

In mathematics we refer to the element of a matrix $\textbf{A}$ that is in the first row, first column as $a_{11}$.  The first index 1 tells us we are interested in the first row, and the second index 1 tells us we are interested in the first column.  The element in row $i$, column $j$ is similarly denoted $a_{ij}$.

**In Python the indexing begins at 0, not at 1**.  If I try and print out the element of the array `A` that is in row 1, column 1 by writing

`print(A[1, 1])`

we get a surprising output

In [3]:
print(A[1, 1])

1


However, the element in row 1 column 1 is equal to 4.  What has gone wrong?

Since the indexing starts at 0, the first row is `A[0]` and the first element of the first row is therefore `A[0, 0]`.  So to output element $a_{11}$ I need to write

`print(A[0, 0])`

In [4]:
print(A[0, 0])

4


If I wish to display the whole of the *third* row I can use

In [5]:
print(A[2])

[3 2 1]


In general, to access the array element in row $i$, column $j$ I need to use

`A[i - 1, j - 1]`

## Creating the Augmented Matrix

The augmented matrix $\left[\,\textbf{A}\,|\,\textbf{I}\,\right]$ has three rows with the matrix $\textbf{A}$ in the first three columns and the identity matrix, $\textbf{I}$, in the final three columns.

The first action in setting up our augmented matrix is to create an empty array of the correct shape, that is an array with three rows and six columns.  We can initialise an array using the `np.zeros` command.  This function will create an array containing the number zero in every element.  

**Subtle point**: The command takes as a *single input argument*; even though we are specifying the number of rows *and* the number of columns.  The required number of rows and columns are specified as a [**tuple**](https://www.tutorialspoint.com/python/python_tuples.htm),

`(rows, columns)`

This is why we see the nested round brackets on the call to `np.zeros` below.  The required code is

`np.zeros((rows, columns))`

In [6]:
Aug = np.zeros((3, 6))
print(Aug)

[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]]


Now we can assign the first three rows and three columns of our array `Aug` to be equal to the array `A`.

We wish to let all the rows and the columns 1, 2, and 3 of `Aug` equal `A`.  To specify that I wish to assign to *all* the rows of `Aug` I use a colon

`Aug[:, `

To specifically assign only columns 1 to 3 to be equal to A I use the following syntax

`Aug[:, 0:3] = `

This looks very odd at first glance.  I wish to make an assignment to the columns indexed by the number 0, 1 and 2.  So why do I write `0:3`?  The answer is that the format for specifying a range of indeces is

`array[start:stop]` and this covers elements from `start` through to `stop-1`.

In effect element `stop` is the first element not to be selected.

So to allocate all rows and the first three columns of `Aug` to be equal to `A` we use 

In [7]:
Aug[:, 0:3] = A
print(Aug)

[[ 4.  2. -1.  0.  0.  0.]
 [ 0.  1.  3.  0.  0.  0.]
 [ 3.  2.  1.  0.  0.  0.]]


The `=` in the assignment only works if the shape of what is being allocated on the left hand side is equal to the shape of what is on the right hand side.  Observe what happens if you incorrectly try

`Aug[:, 1:3] = A`

Now we need to input the identity matrix in columns 4 to 6 of our augmented matrix.  We can do this explicitly in the following way

In [8]:
Aug[0, 3] = 1 # Element (1, 4) etc.
Aug[1, 4] = 1
Aug[2, 5] = 1
print(Aug)

[[ 4.  2. -1.  1.  0.  0.]
 [ 0.  1.  3.  0.  1.  0.]
 [ 3.  2.  1.  0.  0.  1.]]


Although this works, it is not very nice coding.  We have three very similar lines essentially doing the same task.  What would happen if we were working on a $100\times100$ matrix?! When lines of code are very similar we can ask ourselves if it is possible to use a `loop` to execute the task more elegantly. 

We can achieve the same output using either

In [9]:
for i in range(3):
    Aug[i, i + 3] = 1
    
print(Aug)

[[ 4.  2. -1.  1.  0.  0.]
 [ 0.  1.  3.  0.  1.  0.]
 [ 3.  2.  1.  0.  0.  1.]]


or we can make use of the `numpy` function `identity()`.  To create a $3\times3$ identity matrix we use `np.identity(3)`.

In [10]:
Aug[:, 3:6] = np.identity(3)
print(Aug)

[[ 4.  2. -1.  1.  0.  0.]
 [ 0.  1.  3.  0.  1.  0.]
 [ 3.  2.  1.  0.  0.  1.]]


## Carry out Row Operations to Find the Inverse

In this example the first task we will carry out is to choose row operations that will put a zero in the second and third rows, first column of our augmented matrix - shown in red

$$ \textbf{A} = \left(\begin{array}{rrr|rrr} \phantom{-}4 & \phantom{-}2 & -1 & \phantom{-}1 & 0 & 0 \\ \color{red}0 & 1 & 3 & \phantom{-}0 & 1 & 0 \\ \color{red}3 & 2 & 1 & \phantom{-}0 & 0 & 1  \end{array}\right)$$

We see that we are lucky with row 2 as we already have a zero in the first column.

For the third row we can get a zero by using the operation

   R3 $\mapsto$ 4R3 $-$ 3R1
   
which we could code explicitly as

`Aug[2] = 4*Aug[2] - 3*Aug[0]`

but this would be a poor idea - we're not always going to be inverting the matrix given to us in this example.  What we really want to do is to multiply row 3 by the first element of row 1 and then subtract off row 1 multiplied by the first element of row 3; this is much more flexible

`Aug[2] = Aug[0, 0]*Aug[2] - Aug[2, 0]*Aug[0]`

Here we have multiplied row 3, `Aug[2]`, by the first element of row 1, `Aug[0, 0]`, and subtracted off row 1, `Aug[0]` multiplied by the first element of row 3 `Aug[2, 0]`.  

For completeness, let's do the same thing for the second row of `Aug` too, even though in this example there is already a zero where we want one.

In [11]:
Aug[1] = Aug[0, 0]*Aug[1] - Aug[1, 0]*Aug[0] # .. Update row 2
Aug[2] = Aug[0, 0]*Aug[2] - Aug[2, 0]*Aug[0] # .. Update row 3
print(Aug)

[[ 4.  2. -1.  1.  0.  0.]
 [ 0.  4. 12.  0.  4.  0.]
 [ 0.  2.  7. -3.  0.  4.]]


Again, this is not particularly elegant coding - can you see how you could use a loop to improve what we have?

The next sub-task is to put a zero in the second column of the third row, again shown in red

$$ \textbf{A} = \left(\begin{array}{rrr|rrr} 
\phantom{-}4 & \phantom{-}2 & -1 & \phantom{-}1 & 0 & 0 \\ 
           0 & 4 & 12 & \phantom{-}0 & 4 & 0 \\ 
           0 & \color{red}2 & 7 & -3 & 0 & 4  \end{array}\right)$$
           
We can put a zero in row 3 column 2 by carrying out the explicit row operation

   R3 $\mapsto$ 4R3 $-$ 2R2
   
We see that this is really 

   R3 $\mapsto$ $a_{22}$R3 $-$ $a_{32}$R2
   
So we can implement the desired row operation as

In [12]:
Aug[2] = Aug[1, 1]*Aug[2] - Aug[2, 1]*Aug[1]
print(Aug)

[[  4.   2.  -1.   1.   0.   0.]
 [  0.   4.  12.   0.   4.   0.]
 [  0.   0.   4. -12.  -8.  16.]]


The lower left triangle now has zeros everywhere that we need them.  The next sub-task is to consider the upper right triangle and to put zeros in rows 1 and 2, column 3 by choosing suitable row operations.  That is, we need to find appropriate row operations to put zeros where the red elements are

$$ \textbf{A} = \left(\begin{array}{rrr|rrr} 
\phantom{-}4 & \phantom{-}2 & \color{red}-\color{red}1 & \phantom{-}1 & 0 & 0 \\ 
           0 & 4 & \color{red}1\color{red}2 & \phantom{-}0 & 4 & 0 \\ 
           0 & 0 & 4 & -12 & -8 & 16  \end{array}\right)$$

Using the method we have established above, we will carry out

   R1 $\mapsto$ $a_{33}$R1 $-$ $a_{13}$R3
   
   R2 $\mapsto$ $a_{33}$R2 $-$ $a_{23}$R3

In [13]:
Aug[0] = Aug[2, 2]*Aug[0] - Aug[0, 2]*Aug[2] # .. Update row 1
Aug[1] = Aug[2, 2]*Aug[1] - Aug[1, 2]*Aug[2] # .. Update row 2
print(Aug)

[[  16.    8.    0.   -8.   -8.   16.]
 [   0.   16.    0.  144.  112. -192.]
 [   0.    0.    4.  -12.   -8.   16.]]


The final zero we need to set is in the first row second column, here shown in red

$$ \textbf{A} = \left(\begin{array}{rrr|rrr} 
16 & \color{red}8 & 0 & -8 & -8 & 16 \\ 
           0 & 16 & 0 & 144 & 112 & -192 \\ 
           0 & 0 & 4 & -12 & -8 & 16  \end{array}\right)$$
           
The required row operation is

   R1 $\mapsto$ $a_{22}$R1 $-$ $a_{12}$R2
   
and may be implemented via

In [14]:
Aug[0] = Aug[1, 1]*Aug[0] - Aug[0, 1]*Aug[1]
print(Aug)

[[  256.     0.     0. -1280. -1024.  1792.]
 [    0.    16.     0.   144.   112.  -192.]
 [    0.     0.     4.   -12.    -8.    16.]]


The final step is to divide through the elements on the leading diagonal of the augmented matrix to put the identity matrix in the first three rows and columns.  The inverse of $\textbf{A}$ will then be in rows 1 to 3, columns 4 to 6 as we will have found $\left[\,\textbf{I}\,|\,\textbf{A}^{-1}\right]$.

Our final sub-task is therefore to divide through by the elements on the leading diagonal.  When dividing we need to be careful in general, however we will defer consideration of what happens if there is a zero on the leading diagonal to a later session.  For now, let us just divide through each row by the red elements of

$$ \textbf{A} = \left(\begin{array}{rrr|rrr} 
\color{red}2\color{red}5\color{red}6 & 0 & 0 & -1280 & -1024 & 1792 \\ 
           0 & \color{red}1\color{red}6 & 0 & 144 & 112 & -192 \\ 
           0 & 0 & \color{red}4 & -12 & -8 & 16  \end{array}\right)$$
           
Again we could do this explicitly as

R1 $\mapsto$ R1$/a_{11}$

R2 $\mapsto$ R2$/a_{22}$

R3 $\mapsto$ R3$/a_{33}$

But we can see that this task can be more efficiently coded as a `loop` as

In [15]:
for i in range(3):
    Aug[i] = Aug[i]/Aug[i, i]
    
print(Aug)

[[  1.   0.   0.  -5.  -4.   7.]
 [  0.   1.   0.   9.   7. -12.]
 [  0.   0.   1.  -3.  -2.   4.]]


Our solution is now contained in rows 1 to 3 and columns 4 to 6 of our augmented matrix that is stored in the array `Aug`.  We can assign a new $3\times3$ array, let's say `Ainv` to store our solution in.

We assign `Ainv` to be equal to rows 1 to 3 and columns 4 to 6 of our augmented matrix `Aug`.

Again, be careful `Aug[:, 3:6]` refers to *all the rows* of `Aug` and columns 4 to 6 which are indexed by numbers 3, 4, 5 (!).

In [16]:
Ainv = Aug[:, 3:6]
print(Ainv)

[[ -5.  -4.   7.]
 [  9.   7. -12.]
 [ -3.  -2.   4.]]


Hence, we have our solution

$$ \textbf{A} = \left(\begin{array}{rrr} \phantom{-}4 & \phantom{-}2 & -1 \\ 0 & 1 & 3 \\ 3 & 2 & 1 \end{array}\right)
\qquad\Rightarrow\qquad
\textbf{A}^{-1} = \left(\begin{array}{rrr} -5 & -4 & 7 \\ 9 & 7 & -12 \\ -3 & -2 & 4 \end{array}\right)$$

## Checking our Solution

One way of checking our solution is to multiply the input array `A` with our calculated inverse `Ainv` and compare the result with the identity matrix as we should find

$$\textbf{A}\times\textbf{A}^{-1} = \textbf{I}.$$

We can carry out matrix multiplication in a number of different ways.  One way is to use `numpy`'s `matmul` function

In [17]:
print(np.matmul(A, Ainv))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


and we observe that we have the required identity matrix.

## Exercises

- Try the algorithm on a different input matrix of your choice to test that everything still works
- Replace the explicitly coded row operations using loops.  For example we could replace
    
        Aug[1] = Aug[0, 0]*Aug[1] - Aug[1, 0]*Aug[0]
        Aug[2] = Aug[0, 0]*Aug[2] - Aug[2, 0]*Aug[0]

with

    for i in range(2):
        Aug[i + 1] = Aug[0, 0]*Aug[i + 1] - Aug[i + 1, 0]*Aug[0]
    
- When we divide through in the final step we should really check to see that we are not dividing by zero.  What does it mean if we end up in a position where it looks like we would have to divide by zero?
- Can you work out how to use loops in the algorithm so that the same code would work for any sized input array, e.g., a $2\times2$ or a $4\times4$ matrix
- Can you break the algorithm?  What else have we not been careful about?  Try finding the inverse of

$$\textbf{A} = \left(\begin{array}{rrr} \phantom{-}0 & \phantom{-}1 & 4 \\ 4 & 2 & -1 \\ 3 & 2 & 1
\end{array}\right)$$

with

        A = np.array([[0, 1, 4], [4, 2, -1], [3, 2, 1]])
        
This matrix has inverse

$$\textbf{A}^{-1} = \left(\begin{array}{rrr} 4 & 7 & -9 \\ -7 & -12 & 16 \\ 2 & 3 & -4
\end{array}\right)$$

Does your algorithm work on

$$\textbf{A} = \left(\begin{array}{rrr} 2 & -1 & 1 \\ 4 & -2 & 3 \\ -3 & 1 & 1
\end{array}\right) \qquad\Rightarrow\qquad \textbf{A}^{-1} = \left(\begin{array}{rrr} -5 & 2 & -1 \\ -13 & 5 & -2 \\ -2 & 1 & 0
\end{array}\right)$$

and

$$\textbf{A} = \left(\begin{array}{rrr} 2 & 4 & 3 \\ 1 & 2 & 2 \\ 0 & -1 & -2
\end{array}\right) \qquad\Rightarrow\qquad \textbf{A}^{-1} = \left(\begin{array}{rrr} -2 & 5 & 2 \\ 2 & -4 & -1 \\ -1 & 2 & 0
\end{array}\right)$$

## Example Script

In [18]:
#!/usr/bin/env python3
"""This script inverts matrices, where an inverse exists, using the
method of row operations."""

import numpy as np

# User-input array (arbitrary size)
A = np.array([[4, 2, -1], [0, 1, 3], [3, 2, 1]])

m = A.shape[0] # ...................................... Number of rows in A
n = A.shape[1] # ................................... Number of columns in A

# Check matrix is square
if n != m:
    print('Cannot invert a non-square matrix')
else:
    # Create augmented matrix
    Aug = np.zeros((n, 2*n)) # ................ Initialize Augmented matrix
    Aug[:, 0:n] = A # ...................... Put A in left hand side of Aug
    for i in range(n): # ........... Put identity matrix in right hand side
        Aug[i, n + i] = 1

    # Zeros in lower triangle
    for j in range(n - 1):
        Aug[j:n] = Aug[abs(Aug[j:n, j]).argsort()[::-1] + j] # Reverse sort
        for i in range(j + 1, n):
            Aug[i] = Aug[j, j]*Aug[i] - Aug[i, j]*Aug[j]

    # Zeros in upper triangle
    for j in range(n - 1, 0, -1):
        for i in range(j):
            Aug[i] = Aug[j, j]*Aug[i] - Aug[i, j]*Aug[j]

    # Divide through by the elements on the leading diagonal
    SINGULARMATRIX = False # ...... Flag for whether solution exists or not
    for i in range(n):
        if Aug[i, i] != 0: # ............. Check we're not dividing by zero
            Aug[i] = Aug[i]/Aug[i, i]
        else:
            SINGULARMATRIX = True

    # Output solution if possible
    if SINGULARMATRIX is True:
        print('Singular matrix')
    else:
        print('A =\n', A)
        Ainv = Aug[:, n:2*n]
        print('\nInverse found\n',Ainv,'\n')
        print('Error check\nA x A^(-1) =\n',np.matmul(A, Ainv))

A =
 [[ 4  2 -1]
 [ 0  1  3]
 [ 3  2  1]]

Inverse found
 [[ -5.  -4.   7.]
 [  9.   7. -12.]
 [ -3.  -2.   4.]] 

Error check
A x A^(-1) =
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
